futures-util/src/compat/compat01as03.rs

```diff
@@ -95,6 +120,119 @@ impl<St: Stream01> Stream03 for Compat01As03<St> {
+#[must_use = "futures do nothing unless polled"]
+pub struct Compat01As03Sink<S, SinkItem> {
+    pub(crate) inner: Spawn01<S>,
+    pub(crate) buffer: Option<SinkItem>,
+}
+
+impl<S, SinkItem> Unpin for Compat01As03Sink<S, SinkItem> {}
+
+impl<S, SinkItem> Compat01As03Sink<S, SinkItem> {
+    /// Wraps a futures 0.1 Sink object in a futures 0.3-compatible wrapper.
+    pub fn new(inner: S) -> Compat01As03Sink<S, SinkItem> {
+        Compat01As03Sink {
+            inner: spawn01(inner),
+            buffer: None,
+        }
+    }
+
+    fn in_notify<R>(
+        &mut self,
+        lw: &LocalWaker,
+        f: impl FnOnce(&mut S) -> R,
+    ) -> R {
+        let notify = &WakerToHandle(lw.as_waker());
+        self.inner.poll_fn_notify(notify, 0, f)
+    }
+}
+
+impl<S, SinkItem> Stream03 for Compat01As03Sink<S, SinkItem>
+where
+    S: Stream01,
+{
+    type Item = Result<S::Item, S::Error>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> task03::Poll<Option<Self::Item>> {
+        match self.in_notify(lw, |f| f.poll()) {
+            Ok(Async01::Ready(Some(t))) => task03::Poll::Ready(Some(Ok(t))),
+            Ok(Async01::Ready(None)) => task03::Poll::Ready(None),
+            Ok(Async01::NotReady) => task03::Poll::Pending,
+            Err(e) => task03::Poll::Ready(Some(Err(e))),
+        }
+    }
+}
+
+impl<S, SinkItem> Sink03 for Compat01As03Sink<S, SinkItem>
+where
+    S: Sink01<SinkItem = SinkItem>,
+{
+    type SinkItem = SinkItem;
+    type SinkError = S::SinkError;
+
+    fn start_send(
+        mut self: Pin<&mut Self>,
+        item: Self::SinkItem,
+    ) -> Result<(), Self::SinkError> {
+        debug_assert!(self.buffer.is_none());
+        self.buffer = Some(item);
+        Ok(())
+    }
+
+    fn poll_ready(
+        mut self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> task03::Poll<Result<(), Self::SinkError>> {
+        match self.buffer.take() {
+            Some(item) => match self.in_notify(lw, |f| f.start_send(item)) {
+                Ok(AsyncSink01::Ready) => task03::Poll::Ready(Ok(())),
+                Ok(AsyncSink01::NotReady(i)) => {
+                    self.buffer = Some(i);
+                    task03::Poll::Pending
+                }
+                Err(e) => task03::Poll::Ready(Err(e)),
+            },
+            None => task03::Poll::Ready(Ok(())),
+        }
+    }
+
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        lw: &LocalWaker,
+    ) -> task03::Poll<Result<(), Self::SinkError>> {
+        let item = self.buffer.take();
+        match self.in_notify(lw, |f| match item {
+            Some(i) => match f.start_send(i) {
+                Ok(AsyncSink01::Ready) => f.poll_complete().map(|i| (i, None)),
+                Ok(AsyncSink01::NotReady(t)) => {
+                    Ok((Async01::NotReady, Some(t)))
+                }
+                Err(e) => Err(e),
+            },
+            None => Ok((Async01::NotReady, None)),
```

Comment:
This should be `None => f.poll_complete().map(|i| (i, None)),` here as well, right? We don't have a buffered item, but we still need to flush the underlying sink.

Reply:
Ah yup, totally missed that, it should still attempt to `poll_complete` without calling `start_send` like above.